File: iOS/Crisis Connect/Features/Messaging/AuthorityMlsTrustStore.swift
import CryptoKit
import Foundation
import Security

enum AuthorityMlsTrustVerdict: String, Codable, Sendable {
    case first
    case match
    case changed
    case missing
}

struct AuthorityMlsTrustAssessment: Sendable {
    let uid: String
    let verdict: AuthorityMlsTrustVerdict
    let approved: Bool
    let fingerprint: String
    let safetyNumber: String
    let deviceCommitments: [String]
}

enum AuthorityMlsTrust {
    /// Only an already-approved, byte-for-byte matching pin may pass without a new user decision.
    static func approvalCarriesForward(existingApproved: Bool, exactDeviceSetMatch: Bool) -> Bool {
        existingApproved && exactDeviceSetMatch
    }

    static func deviceCommitment(_ record: AuthorityMlsDirectoryRecord) -> String {
        var input = Data()
        appendLengthPrefixed(Data(record.credential.utf8), to: &input)
        appendLengthPrefixed(record.signingPublicKey, to: &input)
        return base64url(Data(SHA256.hash(data: input)))
    }

    static func deviceSetFingerprint(_ commitments: [String]) throws -> String {
        guard !commitments.isEmpty,
              commitments == Array(Set(commitments)).sorted(by: compareUtf8) else {
            throw AuthorityMlsTrustError.invalidDeviceSet
        }
        var input = Data()
        appendLengthPrefixed(Data("cc-authority-mls-device-set:v1".utf8), to: &input)
        for commitment in commitments {
            appendLengthPrefixed(Data(commitment.utf8), to: &input)
        }
        return base64url(Data(SHA256.hash(data: input)))
    }

    static func safetyNumber(_ commitments: [String]) throws -> String {
        let fingerprint = try deviceSetFingerprint(commitments)
        let digest = Data(SHA256.hash(data: Data("cc-authority-mls-safety:v1:\(fingerprint)".utf8)))
        let fields = digest.prefix(16).map { String(format: "%03d", $0) }
        return stride(from: 0, to: fields.count, by: 2)
            .map { fields[$0 ... min($0 + 1, fields.count - 1)].joined(separator: " ") }
            .joined(separator: "  ")
    }

    private static func appendLengthPrefixed(_ field: Data, to output: inout Data) {
        var length = UInt32(field.count).bigEndian
        withUnsafeBytes(of: &length) { output.append(contentsOf: $0) }
        output.append(field)
    }

    private static func base64url(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private static func compareUtf8(_ left: String, _ right: String) -> Bool {
        Array(left.utf8).lexicographicallyPrecedes(Array(right.utf8))
    }
}

/// Approval pins live in a locked-device Keychain item. An authenticated directory row proves only
/// which account session wrote it, not that an already-trusted endpoint authorized the device.
/// First and changed sets therefore remain fail-closed until the exact safety number is approved.
struct AuthorityMlsTrustStore: Sendable {
    // v2 invalidates pins created by the former automatic-approval implementation.
    private static let service = "com.crisisconnect.authority-mls-device-pins-v2"

    func assess(
        conversationId: String,
        uid: String,
        devices: [AuthorityMlsDirectoryRecord]
    ) throws -> AuthorityMlsTrustAssessment {
        guard devices.allSatisfy({ $0.uid == uid }) else {
            throw AuthorityMlsTrustError.invalidDeviceSet
        }
        let commitments = devices.map(AuthorityMlsTrust.deviceCommitment).sorted(by: compareUtf8)
        guard !devices.isEmpty, Set(commitments).count == devices.count else {
            return AuthorityMlsTrustAssessment(
                uid: uid,
                verdict: .missing,
                approved: false,
                fingerprint: "",
                safetyNumber: "",
                deviceCommitments: commitments
            )
        }
        let fingerprint = try AuthorityMlsTrust.deviceSetFingerprint(commitments)
        let safetyNumber = try AuthorityMlsTrust.safetyNumber(commitments)
        let key = recordKey(conversationId: conversationId, uid: uid)
        guard let existing = try load(key) else {
            try save(PinRecord(
                fingerprint: fingerprint,
                deviceCommitments: commitments,
                approved: false,
                verifiedAt: Date().timeIntervalSince1970
            ), key: key)
            return AuthorityMlsTrustAssessment(
                uid: uid,
                verdict: .first,
                approved: false,
                fingerprint: fingerprint,
                safetyNumber: safetyNumber,
                deviceCommitments: commitments
            )
        }
        let matches = existing.fingerprint == fingerprint && existing.deviceCommitments == commitments
        if !matches {
            try save(PinRecord(
                fingerprint: fingerprint,
                deviceCommitments: commitments,
                approved: false,
                verifiedAt: Date().timeIntervalSince1970
            ), key: key)
        }
        return AuthorityMlsTrustAssessment(
            uid: uid,
            verdict: matches ? .match : .changed,
            approved: AuthorityMlsTrust.approvalCarriesForward(
                existingApproved: existing.approved,
                exactDeviceSetMatch: matches
            ),
            fingerprint: fingerprint,
            safetyNumber: safetyNumber,
            deviceCommitments: commitments
        )
    }

    /// Calls fail closed unless the exact current device set was approved previously.
    func verifyExisting(
        conversationId: String,
        uid: String,
        devices: [AuthorityMlsDirectoryRecord]
    ) throws -> AuthorityMlsTrustAssessment {
        try assess(conversationId: conversationId, uid: uid, devices: devices)
    }

    /// Approves only the exact canonical set currently shown to the user.
    func approve(
        conversationId: String,
        uid: String,
        expectedFingerprint: String,
        deviceCommitments: [String]
    ) throws {
        let fingerprint = try AuthorityMlsTrust.deviceSetFingerprint(deviceCommitments)
        guard fingerprint == expectedFingerprint else { throw AuthorityMlsTrustError.invalidDeviceSet }
        let key = recordKey(conversationId: conversationId, uid: uid)
        guard let existing = try load(key),
              existing.fingerprint == fingerprint,
              existing.deviceCommitments == deviceCommitments else {
            throw AuthorityMlsTrustError.storageFailure
        }
        try save(PinRecord(
            fingerprint: fingerprint,
            deviceCommitments: deviceCommitments,
            approved: true,
            verifiedAt: Date().timeIntervalSince1970
        ), key: key)
    }

    private struct PinRecord: Codable {
        let fingerprint: String
        let deviceCommitments: [String]
        let approved: Bool
        let verifiedAt: TimeInterval
    }

    private func recordKey(conversationId: String, uid: String) -> String {
        let digest = SHA256.hash(data: Data("cc-authority-mls-pin:v1:\(conversationId)\u{0}\(uid)".utf8))
        return Data(digest).base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private func load(_ key: String) throws -> PinRecord? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.service,
            kSecAttrAccount as String: key,
            kSecMatchLimit as String: kSecMatchLimitOne,
            kSecReturnData as String: true
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        if status == errSecItemNotFound { return nil }
        guard status == errSecSuccess, let data = item as? Data,
              let record = try? JSONDecoder().decode(PinRecord.self, from: data),
              !record.fingerprint.isEmpty,
              !record.deviceCommitments.isEmpty,
              record.deviceCommitments == record.deviceCommitments.sorted(by: compareUtf8) else {
            throw AuthorityMlsTrustError.storageFailure
        }
        return record
    }

    private func save(_ record: PinRecord, key: String) throws {
        let data = try JSONEncoder().encode(record)
        let lookup: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.service,
            kSecAttrAccount as String: key
        ]
        let update = SecItemUpdate(lookup as CFDictionary, [kSecValueData as String: data] as CFDictionary)
        if update == errSecSuccess { return }
        guard update == errSecItemNotFound else { throw AuthorityMlsTrustError.storageFailure }
        var insert = lookup
        insert[kSecAttrAccessible as String] = kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        insert[kSecValueData as String] = data
        guard SecItemAdd(insert as CFDictionary, nil) == errSecSuccess else {
            throw AuthorityMlsTrustError.storageFailure
        }
    }

    private func compareUtf8(_ left: String, _ right: String) -> Bool {
        Array(left.utf8).lexicographicallyPrecedes(Array(right.utf8))
    }
}

private enum AuthorityMlsTrustError: Error {
    case invalidDeviceSet
    case storageFailure
}

File: Android/app/src/main/java/com/auralis/crisisconnect/messaging/AuthorityMlsTrustStore.kt
package com.auralis.crisisconnect.messaging

import android.content.Context
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.Base64
import org.json.JSONArray
import org.json.JSONObject

enum class AuthorityMlsTrustVerdict { FIRST, MATCH, CHANGED, MISSING }

data class AuthorityMlsTrustAssessment(
    val uid: String,
    val verdict: AuthorityMlsTrustVerdict,
    val approved: Boolean,
    val fingerprint: String,
    val safetyNumber: String,
    val deviceCommitments: List<String>,
)

/**
 * Keeps an authenticated local approval pin for each account's complete MLS device set. Firestore
 * proves only which account session wrote a directory row; first and changed sets therefore remain
 * fail-closed until the user verifies and approves the exact safety number.
 */
class AuthorityMlsTrustStore(context: Context) {
    private val appContext = context.applicationContext

    fun assess(
        conversationId: String,
        uid: String,
        devices: List<AuthorityMlsDirectoryRecord>,
    ): AuthorityMlsTrustAssessment {
        require(devices.all { it.uid == uid }) { "Authority MLS directory was grouped incorrectly." }
        val commitments = devices.map(AuthorityMlsTrust::deviceCommitment).distinct().sorted()
        if (devices.isEmpty() || commitments.size != devices.size) {
            return AuthorityMlsTrustAssessment(uid, AuthorityMlsTrustVerdict.MISSING, false, "", "", commitments)
        }
        val fingerprint = AuthorityMlsTrust.deviceSetFingerprint(commitments)
        val safetyNumber = AuthorityMlsTrust.safetyNumber(commitments)
        val key = recordKey(conversationId, uid)
        val existing = loadPin(key)
        if (existing == null) {
            save(key, PinRecord(fingerprint, commitments, false, System.currentTimeMillis()))
            return AuthorityMlsTrustAssessment(
                uid,
                AuthorityMlsTrustVerdict.FIRST,
                false,
                fingerprint,
                safetyNumber,
                commitments,
            )
        }
        val match = existing.fingerprint == fingerprint && existing.deviceCommitments == commitments
        if (!match) {
            save(key, PinRecord(fingerprint, commitments, false, System.currentTimeMillis()))
        }
        return AuthorityMlsTrustAssessment(
            uid,
            if (match) AuthorityMlsTrustVerdict.MATCH else AuthorityMlsTrustVerdict.CHANGED,
            AuthorityMlsTrust.approvalCarriesForward(existing.approved, match),
            fingerprint,
            safetyNumber,
            commitments,
        )
    }

    /** Calls fail closed unless the exact current device set was approved previously. */
    fun verifyExisting(
        conversationId: String,
        uid: String,
        devices: List<AuthorityMlsDirectoryRecord>,
    ): AuthorityMlsTrustAssessment = assess(conversationId, uid, devices)

    /** Approves only the exact canonical set currently shown to the user. */
    fun approve(
        conversationId: String,
        uid: String,
        expectedFingerprint: String,
        deviceCommitments: List<String>,
    ) {
        require(deviceCommitments.isNotEmpty() && deviceCommitments == deviceCommitments.distinct().sorted()) {
            "Authority MLS approval device set is malformed."
        }
        val fingerprint = AuthorityMlsTrust.deviceSetFingerprint(deviceCommitments)
        require(fingerprint == expectedFingerprint) { "Authority MLS approval fingerprint changed." }
        val key = recordKey(conversationId, uid)
        val existing = loadPin(key) ?: throw SecurityException("Authority MLS approval pin is missing.")
        if (existing.fingerprint != fingerprint || existing.deviceCommitments != deviceCommitments) {
            throw SecurityException("Authority MLS device set changed before approval.")
        }
        save(key, existing.copy(approved = true, verifiedAt = System.currentTimeMillis()))
    }

    private fun save(key: String, record: PinRecord) {
        synchronized(storageLock) {
            val root = loadRoot()
            root.put(key, encodeRecord(record))
            MlsStateVault.saveProtectedData(
                appContext,
                STORAGE_CONTEXT,
                root.toString().toByteArray(StandardCharsets.UTF_8),
            )
        }
    }

    private fun loadPin(key: String): PinRecord? = synchronized(storageLock) {
        val root = loadRoot()
        if (!root.has(key)) null else decodeRecord(root.getString(key))
    }

    private fun loadRoot(): JSONObject {
        val encoded = MlsStateVault.loadProtectedData(appContext, STORAGE_CONTEXT) ?: return JSONObject()
        return JSONObject(String(encoded, StandardCharsets.UTF_8))
    }

    private fun recordKey(conversationId: String, uid: String): String {
        val input = "cc-authority-mls-pin:v1:$conversationId\u0000$uid".toByteArray(StandardCharsets.UTF_8)
        return Base64.getUrlEncoder().withoutPadding().encodeToString(MessageDigest.getInstance("SHA-256").digest(input))
    }

    private data class PinRecord(
        val fingerprint: String,
        val deviceCommitments: List<String>,
        val approved: Boolean,
        val verifiedAt: Long,
    )

    private fun encodeRecord(record: PinRecord): String = JSONObject()
        .put("fingerprint", record.fingerprint)
        .put("deviceCommitments", JSONArray(record.deviceCommitments))
        .put("approved", record.approved)
        .put("verifiedAt", record.verifiedAt)
        .toString()

    private fun decodeRecord(encoded: String): PinRecord {
        val root = JSONObject(encoded)
        val fingerprint = root.getString("fingerprint")
        val array = root.getJSONArray("deviceCommitments")
        val commitments = List(array.length()) { index -> array.getString(index) }
        val verifiedAt = root.getLong("verifiedAt")
        require(
            fingerprint.isNotBlank() &&
                commitments.isNotEmpty() &&
                commitments == commitments.distinct().sorted() &&
                fingerprint == AuthorityMlsTrust.deviceSetFingerprint(commitments) &&
                verifiedAt > 0L
        ) { "Authority MLS device pin is malformed." }
        return PinRecord(fingerprint, commitments, root.getBoolean("approved"), verifiedAt)
    }

    companion object {
        // v2 invalidates pins created by the former automatic-approval implementation.
        private const val STORAGE_CONTEXT = "authority-mls-device-pins:v2"
        private val storageLock = Any()
    }
}

object AuthorityMlsTrust {
    /** Only an already-approved, byte-for-byte matching pin may pass without a new user decision. */
    fun approvalCarriesForward(existingApproved: Boolean, exactDeviceSetMatch: Boolean): Boolean =
        existingApproved && exactDeviceSetMatch

    fun deviceCommitment(record: AuthorityMlsDirectoryRecord): String {
        val digest = MessageDigest.getInstance("SHA-256")
        updateLengthPrefixed(digest, record.credential.toByteArray(StandardCharsets.UTF_8))
        updateLengthPrefixed(digest, record.signingPublicKey)
        return base64url(digest.digest())
    }

    fun deviceSetFingerprint(commitments: List<String>): String {
        require(commitments.isNotEmpty() && commitments == commitments.distinct().sorted())
        val digest = MessageDigest.getInstance("SHA-256")
        updateLengthPrefixed(digest, "cc-authority-mls-device-set:v1".toByteArray(StandardCharsets.UTF_8))
        commitments.forEach { updateLengthPrefixed(digest, it.toByteArray(StandardCharsets.UTF_8)) }
        return base64url(digest.digest())
    }

    fun safetyNumber(commitments: List<String>): String {
        val fingerprint = deviceSetFingerprint(commitments)
        val digest = MessageDigest.getInstance("SHA-256")
            .digest("cc-authority-mls-safety:v1:$fingerprint".toByteArray(StandardCharsets.UTF_8))
        return digest.take(16).map { (it.toInt() and 0xff).toString().padStart(3, '0') }
            .chunked(2).joinToString("  ") { it.joinToString(" ") }
    }

    private fun updateLengthPrefixed(digest: MessageDigest, bytes: ByteArray) {
        digest.update(ByteBuffer.allocate(4).putInt(bytes.size).array())
        digest.update(bytes)
    }

    private fun base64url(bytes: ByteArray): String =
        Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
}
